### PR TITLE
feat(date): support range in calcite-date

### PR DESCRIPTION
**Pull request by rpanichakit9541**
**Related Issue:** #541

## Summary

This adds support for `range` in `calcite-date` component

## Example usage

**Simple**
```html
<calcite-date range ></calcite-date>
```

**Stacked**
```html
<calcite-date range layout="vertical"></calcite-date>
```

**Initial Value**
```html
<calcite-date range layout="vertical" start="2020-10-01" end="2020-10-14"></calcite-date>
```

**With proxies**
```html
<calcite-date range>
   <input slot="start" type="date" value="2020-09-08" >
   <input slot="end" type="date" value="2020-09-23" >
</calcite-date>
```

### CODE REVIEW

**Comment by paulcpederson** [comment]
nice @rpanichakit9541 ! Thanks for whipping this up. 

---

@julio8a @bstifle @macandcheese  I think we need some design decisions from y'all about some of this. There are a couple things that weren't really addressed in the original designs. 



1. When you've selected the start date (but not the end date) and you open the end date dropdown, what should you see? Currently it's the start date selected in blue, this is correct, yeah?
2. Likewise when you have a selected range and you open the start date dropdown, should you see the range (currently just the start is shown)?
3. In the vertical layout version, is it two separate dropdowns?
4. Is there a `no-input` version of the range? Ie just show a calendar and allow range selection?
5. When you select a date we close the wrapper (this is current behavior on single date UI). Would it be better to move focus to the end after they select a start?

Would be good to get a set of 👀 on the padding/borders/drop shadows as well so we know we're meeting the spec.

File: src/components/calcite-date-day/calcite-date-day.scss [hunk 1]
```diff
@@ -40,7 +51,14 @@
 }
 
 :host([scale="m"]) {
-  padding: 4px 4px;
+  .day-v-wrapper {
+    padding-top: 4px;
+    padding-bottom: 4px;
+  }
+  .day-wrapper {
+    padding-left: 4px;
```
Comment (paulcpederson) [inline src/components/calcite-date-day/calcite-date-day.scss:59]: We've just recently added tailwind to this repo, so I think we could get these values from the spacing scale with padding modifiers and and the `@apply` directive like:

```

:host([scale="s"]) {
  .day-v-wrapper {
    padding-top: 2px;
    padding-bottom: 2px;
  }
  .day-wrapper {
    @apply px-0;
  }
  .day {
    height: 27px;
    width: 27px;
    @apply text--2;
  }
}

:host([scale="m"]) {
  .day-v-wrapper {
    @apply py-1;
  }
  .day-wrapper {
    @apply px-1;
  }
  .day {
    height: 33px;
    width: 33px;
    @apply text--1;
  }
}

:host([scale="l"]) {
  .day-v-wrapper {
    @apply py-1;
  }
  .day-wrapper {
    @apply px-1;
  }
  .day {
    height: 43px;
    width: 43px;
    @apply text-0;
  }
}
```

This hasn't made it to our conventions doc, but if you want to be a trail blazer you could play around with it.

I'm also unsure as to how you would express `2px` or the other units in tailwind-speak. It's pretty new to me...

File: src/components/calcite-date-month/calcite-date-month.tsx [hunk 2]
```diff
@@ -301,4 +274,59 @@ export class CalciteDateMonth {
...
+    active: boolean,
+    day: number,
+    dir: string,
+    date: Date,
+    currentMonth?: boolean,
+    ref?: boolean
+  ) {
+    const props = {
+      active,
+      ...(currentMonth && { currentMonth: true }),
```
Comment (paulcpederson) [inline src/components/calcite-date-month/calcite-date-month.tsx:305]: did this not work as

```
active,
currentMonth, 
...
```

I _think_ stencil will conditionally add/remove boolean props if what you pass it evaluates to false, but I've never passed a spread object like this, so I may be way off.

File: src/components/calcite-date-month/calcite-date-month.tsx [hunk 3]
```diff
@@ -301,4 +274,59 @@ export class CalciteDateMonth {
...
+      day,
+      dir,
+      disabled: !inRange(date, this.min, this.max),
+      startOfRange:
```
Comment (paulcpederson) [inline src/components/calcite-date-month/calcite-date-month.tsx:309]: it may read a bit better to just make this conditional a method on this component like `this.isStartOfRange`.

File: src/components/calcite-date/calcite-date.tsx [hunk 4]
```diff
@@ -127,68 +183,83 @@ export class CalciteDate {
...
+                  type="text"
+                  value={formattedDate}
+                />
+              )}
+              {(!this.range || (this.focusedInput === "start" && this.layout !== "vertical")) &&
+                this.renderCalendar(activeDate, dir, maxDate, minDate, date, endDate)}
+            </div>
+            {this.range && this.layout === "horizontal" && (
+              <div class="horizontal-arrow-container">
+                <calcite-icon icon={dir === "ltr" ? `arrow-right` : `arrow-left`} scale="s" />
```
Comment (paulcpederson) [inline src/components/calcite-date/calcite-date.tsx:234]: you can also do `icon="arrow-right" mirrored` (although that prop is about to change to be called `flip-rtl`.

**Review by paulcpederson — APPROVED** [review]

**Comment by paulcpederson** [comment]
Had a couple comments, but nothing major. Great work 💯

File: src/components/calcite-date/calcite-date.tsx (hunk 4 above)
Comment (rpanichakit9541) [reply src/components/calcite-date/calcite-date.tsx:234]: oh that's cool.  Didn't know that.  If I use it, when the prop name change will it automatically refactor my usage too or should I wait to use it after the change.

File: src/components/calcite-date-month/calcite-date-month.tsx (hunk 2 above)
Comment (rpanichakit9541) [reply src/components/calcite-date-month/calcite-date-month.tsx]: You are right.  This was overkill, I've updated it.

File: src/components/calcite-date-day/calcite-date-day.scss (hunk 1 above)
Comment (rpanichakit9541) [reply src/components/calcite-date-day/calcite-date-day.scss:59]: Awesome, let me take a look at this.

**Comment by bstifle** [comment]
> nice @rpanichakit9541 ! Thanks for whipping this up.
> 
> @julio8a @bstifle @macandcheese I think we need some design decisions from y'all about some of this. There are a couple things that weren't really addressed in the original designs.
> 
> 
> 
> 1. When you've selected the start date (but not the end date) and you open the end date dropdown, what should you see? Currently it's the start date selected in blue, this is correct, yeah?
> 2. Likewise when you have a selected range and you open the start date dropdown, should you see the range (currently just the start is shown)?
> 3. In the vertical layout version, is it two separate dropdowns?
> 4. Is there a `no-input` version of the range? Ie just show a calendar and allow range selection?
> 5. When you select a date we close the wrapper (this is current behavior on single date UI). Would it be better to move focus to the end after they select a start?
> 
> Would be good to get a set of 👀 on the padding/borders/drop shadows as well so we know we're meeting the spec.



1. yes
2. yes
3. would be dope to do 1 dropdown
4. i could see that. do we have a case for it?
5. yes

**Comment by macandcheese** [comment]
Awesome @rpanichakit9541. Yeah I can see a "no input" version just like for the regular one, for some large single - step workflows around date setting, or scheduling, with no shadow in that case correct?

**Comment by rpanichakit9541** [comment]
For the "no-input" version, when the user click on a date how do we determine if they are wanting to change the start or end value?

File: src/components/calcite-date-month-header/calcite-date-month-header.tsx [hunk 5]
```diff
@@ -90,14 +90,14 @@ export class CalciteDateMonthHeader {
...
             aria-label={this.intlNextMonth}
             class="chevron"
             href="#"
             onClick={(e) => this.handleArrowClick(e, prevMonthDate)}
             onKeyDown={(e) => this.handleKeyDown(e, prevMonthDate)}
             role="button"
-            tabindex="0"
+            tabindex={prevMonthDate.getMonth() === activeMonth ? -1 : 0}
           >
             <calcite-icon dir={dir} icon="chevron-left" mirrored scale={iconScale} />
```
Comment (macandcheese) [inline src/components/calcite-date-month-header/calcite-date-month-header.tsx:102]: As @paulcpederson mentioned - this just recently changed to `flip-rtl` in this PR that was merged: https://github.com/Esri/calcite-components/pull/1132

File: src/components/calcite-date/calcite-date.tsx (hunk 4 above)
Comment (rpanichakit9541) [reply src/components/calcite-date/calcite-date.tsx:234]: Just saw @macandcheese 's comment that the prop name changed was merged, so I'll update this to use `flip-rtl` .

File: src/components/calcite-date-day/calcite-date-day.scss (hunk 1 above)
Comment (rpanichakit9541) [reply src/components/calcite-date-day/calcite-date-day.scss:59]: @paulcpederson - a general question on this.  For things like 
```css
:host {
  display: inline-block;
  vertical-align: top;
  width: 100%;
  position: relative;
  overflow: visible;
}
```
Do we want to replace it completely with tailwind like below, or are we ok with what we had?

```
:host {
  @apply inline-block align-top w-full relative overflow-visible;
}
```

**Comment by paulcpederson** [comment]
@rpanichakit9541 regarding "which you are setting" there are two approaches to this:

1. Once you have a range, a click starts a new range selection (ie, the other date is cleared)
2. Once you have a range, a click updates the closest of the two (ie, closer to start edits start date)

https://ember-power-calendar.com/docs/range-selection has a pretty good summary of these, the `2.` option would be their `proximitySelection` flag.

**Comment by bstifle** [comment]
> @rpanichakit9541 regarding "which you are setting" there are two approaches to this:
> 
> 1. Once you have a range, a click starts a new range selection (ie, the other date is cleared)
> 2. Once you have a range, a click updates the closest of the two (ie, closer to start edits start date)
> 
> https://ember-power-calendar.com/docs/range-selection has a pretty good summary of these, the `2.` option would be their `proximitySelection` flag.


I think 2. is the better solution, but 1. is a good fallback

**Comment by bstifle** [comment]
@rpanichakit9541 @paulcpederson 

1. I think we need a hover state for the range. See design:
image

2.Also, thinking it is a little odd when switching months inside the range. I don't think the focus should "follow you".

E.g. if you have the 28th selected, and then click to the next month, the 28th of that month is automatically selected. Kind of unexpected UX

image 
image

**Comment by bstifle** [comment]
@rpanichakit9541 @paulcpederson

Last things sorry, just noticed:

3. The type color on the selected range dates (non the start and finish) should be ui-text-1 (#151515)
4. The light blue fill on the dates should be ui-blue-1 with a opacity set to 15%

image

**Comment by bstifle** [comment]
@rpanichakit9541

5. Annnddd some more. For the dividing border, it appears to be doubled up from the inputs and is displaying 2px.
image


Can we get it closer to this design, with the padding on the right as well?:
image

**Comment by bstifle** [comment]
@rpanichakit9541

For dark theme, please make the dates in between the range selected ui-blue-1-dark at 15% opacity. And the type set to ui-text-1-dark.


image

**Comment by paulcpederson** [comment]
> The light blue fill on the dates should be ui-blue-1 with a opacity set to 15%

@bstifle is this in our palette somewhere? Currently using `--calcite-ui-foreground-current` which is `$h-bb-010` so if we change the color here, IMO it should change everywhere.

/cc @asangma

**Comment by bstifle** [comment]
> > The light blue fill on the dates should be ui-blue-1 with a opacity set to 15%
> 
> @bstifle is this in our palette somewhere? Currently using `--calcite-ui-foreground-current` which is `$h-bb-010` so if we change the color here, IMO it should change everywhere.
> 
> /cc @asangma

Hmm? ui-blue-1 is the variable for #007ac2

**Comment by paulcpederson** [comment]
@bstifle You can't set opacity of a background color in CSS using a hexadecimal. We would have to use RGBA or HSLA to set this. If we use `opacity` the entire element will be `.15` opacity (including the number). Much easier to set the background color with a... color. 

When you make a color .15 opacity over white it's a different color. If we allowed any color from our vars to be used at every opacity, we would have infinite colors... 

TLDR: do we have a non-alpha-channel version of this color available to use?

**Comment by bstifle** [comment]
@rpanichakit9541 let's just use

rgba(0,122,194, 0.15) for light theme

rgba(0,160,255, 0.15) for dark theme

**Comment by bstifle** [comment]
seeing about adding these colors as variables so that we can theme them later on down the road

https://github.com/Esri/calcite-components/issues/1166

**Comment by macandcheese** [comment]
To the non-alpha request, doesn't MVB have a "light blue" in use @asangma? Could use that here since I feel like it's close.

**Comment by bstifle** [comment]
> To the non-alpha request, doesn't MVB have a "light blue" in use @asangma? Could use that here since I feel like it's close.

Yep, see https://github.com/Esri/calcite-components/issues/1166

**Comment by rpanichakit9541** [comment]
@paulcpederson @macandcheese @bstifle - I've made some changes per the first round of feedback.  Can you guys take another look when you get a chance?  Let me know if I missed anything or if anything else needs to be changed/improved.  Thank you!

**Comment by paulcpederson** [comment]
@rpanichakit9541 hmmm looks like we've got build failure shenanigans. It looks good to me, but would be good to get the tests passing. Let me know if you need help and I can try to help debug.

**Comment by rpanichakit9541** [comment]
@paulcpederson - Thanks yeah it looks like a timing issue on `expect(changedEvent).toHaveReceivedEventTimes(1);`  Not sure if there's a way to ensure that the event fires after the "space" press before executing that line.  When I run the test locally it passes about 50% of the time, I've tried adding `setTimeout` before it and it does help it pass most of the time, but not on the build server.  Have you encounter something similar to this before?

**Comment by paulcpederson** [comment]
I think I had this issue as well. Wondering if I ever tried a `page.waitForChanges` ?

**Comment by bstifle** [comment]
@rpanichakit9541 heyo! nice changes, just a couple comments to tighten it all up:

Screen Shot 2020-10-26 at 9 15 49 AM
1. When range is selected, the gray hover state needs to extend behind the blue selection

 
Screen Shot 2020-10-26 at 9 16 16 AM
2. The left and right edges of the gray hover state need to be rounded when wrapping


Screen Shot 2020-10-26 at 9 16 49 AM
3. Light theme, active range numbers in between selected should be ui-text-1 (#151515)


Screen Shot 2020-10-26 at 9 34 51 AM
4. Dark theme, active range numbers in between selected should be ui-text-1-dark (#ffffff)


image
5. Dark theme, active color should be m-bb-090 (#214155)


image
6. I dont know why, but sometimes, after selecting range, and then editing the start date, the end date resets and is no longer selected

**Comment by paulcpederson** [comment]
@rpanichakit9541 sorry I totally made your life difficult fixing that other date issue... 😞 🙏

**Comment by rpanichakit9541** [comment]
@paulcpederson - haha no problem. I'll start looking at resolving the conflicts.

**Comment by rpanichakit9541** [comment]
@paulcpederson - I noticed in your last fix you removed this function below.
```
 @Watch("value")
  valueWatcher(value: string): void {
    this.setValueAsDate(value);
  }
```
I believe this caused the "Set Value" example on the test page to stop working .  `document.getElementById("dynamic-value").value = "2020-10-02" no longer correctly update the component.

If I add the watch back in, it works again.  Should I do that or is there a better solution?

**Comment by paulcpederson** [comment]
@rpanichakit9541 ok, yeah I probably broke that. We just need to confirm that in other timezones it doesn't cause a render loop when we add that back in. Easiest way to check is to create a chrome profile and then launch a new chrome instance setting the timezone:

```
mkdir ~/chrome-profile
TZ='Asia/Shanghai' open -na "Google Chrome" --args "--user-data-dir=$HOME/chrome-profile"
```

https://github.com/Esri/calcite-components/issues/1111#issuecomment-710701159

**Comment by paulcpederson** [comment]
Or if you're on windows I think you can set the time zone on your OS?

**Comment by paulcpederson** [comment]
@rpanichakit9541 testing locally on your branch, seems like adding:

```
  @Watch("value")
  valueWatcher(value: string): void {
    this.valueAsDate = dateFromISO(value);
  }
```

works fine and doesn't cause a regression.

**Comment by rpanichakit9541** [comment]
@paulcpederson - Thank you for checking! I'll add it back in then.

**Comment by rpanichakit9541** [comment]
@bstifle - I've made updates according to your comment https://github.com/Esri/calcite-components/pull/1136#issuecomment-716717094 .  The only thing I did not do was "5. Dark theme, active color should be m-bb-090 (#214155)" .  Currently it's using `--calcite-ui-blue-1` so it automatically get the right color depending on the theme.  Do we have an example in another component where we deviate and set a specific color on dark theme like that?

@paulcpederson - Can you also do another review when you have some time? Thank you.

**Comment by bstifle** [comment]
@rpanichakit9541 

Looking pretty good!

"Do we have an example in another component where we deviate and set a specific color on dark theme like that?"

Yeah it's weird, but there are no other instances like this. so it needs to be a one off. Please apply that #214155 on dark theme.

One visual bug I found: if you have a range over a couple months long, when you go to edit, the selected end date appears as idle on all months.

image
image

**Comment by rpanichakit9541** [comment]
@bstifle - I've changed the dark theme active range color to #214155 as requested and fixed the visual bug that you pointed out.  Let me know if there are other issues.

**Comment by bstifle** [comment]
@rpanichakit9541 nice! thank you!

Ok so I guess its showing this now, for future months days that are the same number as selected days in the range it's showing the gray hover state. same if you go backwards in month with the first selection

image
image

**Comment by rpanichakit9541** [comment]
@bstifle - ok the gray state should now be gone from both cases.

**Comment by paulcpederson** [comment]
@bstifle isn't that gray the active indicator? When you tab into the calendar, that day will receive focus, and arrow keys let you edit it, space/enter let you select it. If we're removing this indicator she should probably do it on single select date as well. But Maybe I'm misunderstanding...

@rpanichakit9541 this looks good to me, thanks for sticking with it!

**Review by bstifle — APPROVED** [review]
Looks super good!!!!!! Nice work!!!

**Comment by bstifle** [comment]
@paulcpederson isn't that gray the active indicator?


no we use blue for that https://www.youtube.com/watch?v=68ugkg9RePc

**Comment by rpanichakit9541** [comment]
@paulcpederson @macandcheese @bstifle - Anything else left to be done on this PR before it can be merged?
